### DISCUSSION TITLE
Can I please get an idea about the file structure pls.

### DISCUSSION THREAD

**Discussion by DIVYANSGupta** [requarks/wiki#6872] [Help / Questions] [answered]
I tried to run the wiki app through the docker image, but I want to run the vue app service separately on port(3001) and node ap service separately on port(3000). 
Please help me with file structure to achieve my goal.

**Comment by DIVYANSGupta** [comment]
Hey, @NGPixel, could you please help me  to know the file structure.

**Comment by NGPixel** [comment]
Use the app the way it's intended instead of trying to split it for no reason?

**Reply by DIVYANSGupta** [reply]
Hey @NGPixel 
Actually, I wanted to add more vue components to wiki app to page (/).
I tried modifying the pug file directly but it's consuming more time.

**Reply by DIVYANSGupta** [reply]
hey @NGPixel Can we discuss on this further, any other resource whom i can contact to get information on this

**Reply by NGPixel** [reply, marked as answer]
Unless you have a **very specific** question I can answer, no. I don't have time to guide you through altering the project to do whatever you're trying to do, which still makes no sense to me. The pug files have nothing to do with vue components or client code.

The docs already have the necessary info to get started in modifying client code:
https://docs.requarks.io/dev
https://docs.requarks.io/dev/themes

**Reply by DIVYANSGupta** [reply]
Hey @NGPixel , I understand your point. 
On the home screen page I want to move the author(last edited by card) and sharing option to the nav bar (black section : user-profile). How can I achieve this? also I want to add/update the UI components. Whatever the changes I do in my vue file doesn't reflect on browser, Please assist me on this.
![image](https://github.com/requarks/wiki/assets/148762020/f92e9ae9-81b3-4dbc-aac3-7344b14eb896)

**Reply by DIVYANSGupta** [reply]
hey @NGPixel  could you help me with it?

**Reply by NGPixel** [reply]
1. **STOP tagging me with multiple replies.** This is annoying. You posted only 4 hours ago and somehow you already expect a response?? This isn't express dev support. It can take days to get a reply IF someone knows the answer. Constantly nagging people will only result in getting you blocked from this repo.

2. I already pointed you to the relevant docs, which clearly explain that you must build the client files to see changes. Just modifying the .vue files won't do anything.

**Comment by rrg92** [comment]
Changing the client involves extensive knowledge in Vue.js, JavaScript, Webpack, Pug, CSS (including SCSS), and Node.js. It's not as simple as just modifying a `.html` file and requires a considerable amount of time.

If you wish to customize, I suggest you start by learning these technologies if you're not already familiar with them.

If you are, then try following the building documentation provided by NgPixel.

**General Steps to Customize the Interface:**

1. **Identify the Source Files:** Identify the source files of the screen you want to edit.
2. **Understand the Building Process:** The interface seen in the browser is generated by building the client via Webpack. This task is complex and requires studying the wiki structure. However, once you understand it, it becomes easier to make more edits. This initial learning phase is crucial.
3. **Editing Files:** For example, if you want to change something in the editor, you might need to modify files like `client/components/editor.vue` or `client/components/editor/ckeditor.vue`, depending on your needs. Most file names are self-explanatory, but a solid understanding of Vue.js and how compoennts works is vital.

These changes can range from simple HTML modifications to more complex alterations, depending on your requirements. You also need to be cautious and conduct extensive testing to ensure that your changes do not break your wiki. In-depth knowledge of Vue.js 2, along with Pug, is crucial here, as all pages are Vue components.

After editing, you must build the client and replace the files on your production server. This task can be complex too, and thorough testing is essential. Attempting it without testing can break your wiki.

**Alternative Approach Using JavaScript:**

If you prefer, you can try making edits just in JavaScript. This approach can be complex initially, but give opportunity to edit without toucing vue. For example, create an endpoint in Node.js (e.g., `server/controllers/common.js`) that returns a static file called `mycustom.js` and add this path to `page.pug`. This requires understanding Node.js, Express, and Pug. Then, you can use `mycustom.js` to write functions that change the page after the DOM loads, using the custom `window.boot` provided by the wiki. You can change many aspects at runtime using this method without compiling Vue, but note that this is not considered best practice for that project.

**Important Reminder:**

Remember, every customization is at your own risk, and be sure to backup your data before replacing any files.



**Understanding Pug Files vs. Pug Language in NGPixel's Context**

As NGPixel mentioned, it's important to note that editing `.pug` files is not always necessary. These files, typically server-side, essentially return "HTML" code to be rendered in the browser, functioning similarly to traditional HTML/HTTP protocol applications. My suggestion to edit these `.pug` files is merely an alternative approach. For instance, it can be useful if you want to add a new `<script>` tag to be loaded on the client side without altering Vue components.

However, it's crucial to distinguish between 'pug files' and the 'pug language':

- **Pug Files:** When referring to ".pug files", it's about the files located in the `server/views` directory. These are primarily used for rendering the initial HTML structure sent to the client.

- **Pug Language:** On the other hand, when mentioning "pug language", it pertains to the usage of the Pug template syntax within many `.vue` files under the `client` folder. This involves a deeper integration with Vue components, where Pug provides a templating syntax for structuring the component's markup.

**Key Takeaway:**
Understanding this distinction is essential for effectively navigating and customizing the application. While both involve Pug, their roles and implications in the project's architecture differ significantly.

### A Final Important Note: My Experience and Gratitude

I've been deeply involved in making significant customizations to this project, which has required extensive study and understanding of various technologies. I've invested a lot of time in learning these languages, testing them, and then understanding how Wiki.js incorporates them. Only after this comprehensive learning process was I able to start customizing for my own needs.

I fully understand that NGPixel, or any other developer of Wiki.js, has their own time commitments and personal projects. They cannot dedicate all their time to simply guiding us through customizations. The work already done on Wiki.js is a substantial contribution to the community, and we should be patient and try to help in any way we can. I plan, in due time, to commit back all the changes I've made, making them available for everyone's use, thereby returning the favor to the Wiki.js community.

**A Plea for Patience and Respect:**
Please, let's not stress the developers or maintainers. With the little time I've been able to dedicate to customizing Wiki.js, I know how challenging it is to keep everything working, tested, etc. Patience is key. We must appreciate the hard work and dedication of those who contribute to projects like Wiki.js, understanding the complex balance they maintain between their personal commitments and their contributions to open-source projects.
